upup/pkg/fi/cloudup/awstasks/warmpool.go

```diff
@@ -0,0 +1,130 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awstasks
+
+import (
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"k8s.io/klog/v2"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/awsup"
+	"k8s.io/kops/upup/pkg/fi/cloudup/cloudformation"
+	"k8s.io/kops/upup/pkg/fi/cloudup/terraform"
+)
+
+// AutoscalingGroup provdes the definition for a autoscaling group in aws
+// +kops:fitask
+type WarmPool struct {
+	// Name is the name of the ASG
+	Name *string
+	// Lifecycle is the resource lifecycle
+	Lifecycle *fi.Lifecycle
+
+	Enabled *bool
+	// MaxSize is the max number of nodes in the warm pool
+	MaxSize *int64
+	// MinSize is the smallest number of nodes in the warm pool
+	MinSize *int64
+
+	AutoscalingGroup *AutoscalingGroup
+}
+
+// Find is used to discover the ASG in the cloud provider
+func (e *WarmPool) Find(c *fi.Context) (*WarmPool, error) {
+	cloud := c.Cloud.(awsup.AWSCloud)
+	svc := cloud.Autoscaling()
+	warmPool, err := svc.DescribeWarmPool(&autoscaling.DescribeWarmPoolInput{
+		AutoScalingGroupName: e.Name,
+	})
+	if err != nil {
+		return nil, err
+	}
+	if warmPool.WarmPoolConfiguration == nil {
+		return &WarmPool{
+			Name:    e.Name,
+			Enabled: fi.Bool(false),
+		}, nil
+	}
+
+	actual := &WarmPool{
+		Name:    e.Name,
+		Enabled: fi.Bool(true),
+		MaxSize: warmPool.WarmPoolConfiguration.MaxGroupPreparedCapacity,
+		MinSize: warmPool.WarmPoolConfiguration.MinSize,
+	}
+	return actual, nil
+}
+
+func (e *WarmPool) Run(c *fi.Context) error {
+	return fi.DefaultDeltaRunMethod(e, c)
+}
+
+func (*WarmPool) CheckChanges(a, e, changes *WarmPool) error {
+	return nil
+}
+
+func (*WarmPool) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *WarmPool) error {
+
+	svc := t.Cloud.Autoscaling()
+	if changes != nil {
+		if fi.BoolValue(e.Enabled) {
+			minSize := e.MinSize
+			if minSize == nil {
+				minSize = fi.Int64(-1)
+			}
+			maxSize := e.MaxSize
+			if maxSize == nil {
+				maxSize = fi.Int64(-1)
+			}
+			request := &autoscaling.PutWarmPoolInput{
+				AutoScalingGroupName:     e.Name,
+				MaxGroupPreparedCapacity: maxSize,
```

Comment:
In AWS, the max prepared capacity can be no less than the minsize of the warm pool. We should either enforce that through API validation or silently increase the max prepared capacity to be at least the minsize. The fact that the max prepared capacity defaults to the IG maxsize is a complication.

Reply:
Although AWS allows the maxPrepared to be smaller than the IG minSize, such a configuration is nonsensical. Perhaps we should add an API validation for that.

Reply:
I am hesitant to restrict our API more than the AWS API. But if users could get burned by this, I think it makes sense.